transcendence-app/src/shared/db/postgres/utils.ts

```diff
@@ -36,28 +36,31 @@ export const makeQuery = async <T>(
   }
 };
 
-export function insertWithClient<T extends Record<string, any>>(
-  client: PoolClient,
-  table: table,
-  entity: T,
-) {
-  const { cols, params, values } = entityQueryMapper(entity);
-  const text = `INSERT INTO ${table}(${cols.join(', ')}) VALUES (${params.join(
-    ',',
-  )}) RETURNING *;`;
-  return client.query(text, values);
-}
-export function updateByIdWithClient<T extends Record<string, any>>(
-  client: PoolClient,
-  table: table,
-  id: string,
-  chatroom: T,
-) {
-  const { cols, values } = entityQueryMapper(chatroom);
-  const colsToUpdate = cols.map((col, i) => `${col}=$${i + 2}`).join(',');
-  const text = `UPDATE ${table} SET ${colsToUpdate} WHERE "id"=$1 RETURNING *;`;
-  return client.query(text, [id, ...values]);
-}
+export const makeTransactionalQuery = async <T>(
+  pool: PostgresPool,
+  callback: (client: PoolClient) => Promise<T | null>,
+) => {
+  let data: T | null = null;
+  try {
+    const client = await pool.connect();
+    try {
+      await client.query('BEGIN');
+      data = await callback(client);
+      await client.query('COMMIT');
+    } catch (err) {
+      await client.query('ROLLBACK');
+      throw err;
+    } finally {
+      client.release();
```

Comment:
BTW, do we go into this block if we enter the above catch, and rethrow? I would expect not, but maybe we should release all the time anyway? Just wondering.

Reply:
> finallyStatements
Statements that are executed before control flow exits the try...catch...finally construct. These statements execute regardless of whether an exception was thrown or caught.


According to MDN, yes.

Reply:
If we return in the above catch, as per my suggestion, are we still hitting the finally?
That was my concern.

Reply:
Oh. Yes, the finally block is executed even if we return from the try or the catch block.
I've changed it to return after the try/catch/block because it was confusing for me :sweat_smile: